### PR TITLE
Add a pass to vectorize Linalg ops.

### PR DESCRIPTION
**Pull request by hanhanW**
The convert op is the only one not covered, so we have an extra flag to
enable the pass. It needs an upstream fix in MLIR.

A Linalg generic op will be vectorized only if:
  1) All the memory access is contiguous.
  2) The shape of the last dim is a multiple of 4.
  3) All the operands have only one use, which is the generic op.

This pass can also be extended to create large vectors. In the patch, it
only consider small vectors like 4xf32/4xi32.

### CODE REVIEW

**Comment by asaadaldien** [comment]
Cool! This pass is useful to be shared between LLVM and SPIRV lowering, 
How about moving it to new directory 'Conversion/LinalgToVector/...' ?

File: iree/compiler/Conversion/LinalgToSPIRV/VectorizeLinalgOps.cpp
Comment (asaadaldien) [inline iree/compiler/Conversion/LinalgToSPIRV/VectorizeLinalgOps.cpp:45]: Its better have this as a flag e.g `*-linalg-vector-size` then drive the bit size from it (just * 8).

File: iree/compiler/Conversion/LinalgToSPIRV/VectorizeLinalgOps.cpp
Comment (asaadaldien) [inline iree/compiler/Conversion/LinalgToSPIRV/VectorizeLinalgOps.cpp:61]: This can be set from the vector size flag

File: iree/compiler/Conversion/LinalgToSPIRV/VectorizeLinalgOps.cpp
Comment (asaadaldien) [inline iree/compiler/Conversion/LinalgToSPIRV/VectorizeLinalgOps.cpp:79]: Another idea is to rewrite it as a slpat constant of vector size. to avoid adding  a broadcast op

File: iree/compiler/Conversion/LinalgToSPIRV/VectorizeLinalgOps.cpp
Comment (asaadaldien) [inline iree/compiler/Conversion/LinalgToSPIRV/VectorizeLinalgOps.cpp:240]: Why not a function pass ?

File: iree/compiler/Conversion/LinalgToSPIRV/VectorizeLinalgOps.cpp
Comment (asaadaldien) [inline iree/compiler/Conversion/LinalgToSPIRV/VectorizeLinalgOps.cpp:179]: We also need to check inner most iterator is parallel

File: iree/compiler/Conversion/LinalgToSPIRV/VectorizeLinalgOps.cpp
Comment (hanhanW) [reply iree/compiler/Conversion/LinalgToSPIRV/VectorizeLinalgOps.cpp]: It is already doing this way. I think we still can add a broadcast op and apply canonicalize pass or create it using createOrFold, so it will still go away. This is not implemented yet.

File: iree/compiler/Conversion/LinalgToSPIRV/VectorizeLinalgOps.cpp
Comment (hanhanW) [reply iree/compiler/Conversion/LinalgToSPIRV/VectorizeLinalgOps.cpp]: It expects to operate on element-wise pass, so check if all the loops are parallel.

**Comment by hanhanW** [comment]
> Cool! This pass is useful to be shared between LLVM and SPIRV lowering,
> How about moving it to new directory 'Conversion/LinalgToVector/...' ?

Done!

It seems to need more work to have `*-linalg-vector-size`, could I do it in a follow change?

**Comment by asaadaldien** [comment]
> > Cool! This pass is useful to be shared between LLVM and SPIRV lowering,
> > How about moving it to new directory 'Conversion/LinalgToVector/...' ?
> 
> Done!
> 
> It seems to need more work to have `*-linalg-vector-size`, could I do it in a follow change?

SGTM! we can land this and iterate. thanks for starting this pass.

**Review by asaadaldien — APPROVED** [review]

File: iree/compiler/Conversion/LinalgToVector/VectorizeLinalgOps.cpp
Comment (MaheshRavishankar) [inline iree/compiler/Conversion/LinalgToVector/VectorizeLinalgOps.cpp:33]: Maybe move this as a utility method to MLIR?

File: iree/compiler/Conversion/LinalgToVector/VectorizeLinalgOps.cpp
Comment (MaheshRavishankar) [inline iree/compiler/Conversion/LinalgToVector/VectorizeLinalgOps.cpp:173]: I dont think this is needed. It should just pass through fine.

File: iree/compiler/Conversion/LinalgToVector/VectorizeLinalgOps.cpp
Comment (MaheshRavishankar) [inline iree/compiler/Conversion/LinalgToVector/VectorizeLinalgOps.cpp:189]: You can just check this way I think

```
AffineDimExpr innerMostExpr = map.getResult().back().dyn_cast<AffineDimExpr>();
if (!innerMostExpr || innerMostExpr.getPosition() != map.getNumDims() - 1) {
  return failure();
}
```

File: iree/compiler/Conversion/LinalgToVector/VectorizeLinalgOps.cpp
Comment (MaheshRavishankar) [inline iree/compiler/Conversion/LinalgToVector/VectorizeLinalgOps.cpp:201]: Why not handle `iree.placeholder` ops separately in a pattern?

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]

File: iree/compiler/Conversion/LinalgToVector/VectorizeLinalgOps.cpp
Comment (hanhanW) [reply iree/compiler/Conversion/LinalgToVector/VectorizeLinalgOps.cpp:173]: This is a tricky one that I found in a model. Through several passes (fusion, canonicalize, etc) we would get a snippet in the dispatch function.

```
%cst = constant 1.0 : f32
linalg.generic { ... } %output_buf
^bb(...):
  linalg.yield %cst
```

There are two other options:

1) Have another pass to rewrite Linalg ops to named ops.
2) Handle this in yield op pattern.

The first approach looks better to me. This is just a workaround to make us be able to run a E2E model.

File: iree/compiler/Conversion/LinalgToVector/VectorizeLinalgOps.cpp
Comment (hanhanW) [reply iree/compiler/Conversion/LinalgToVector/VectorizeLinalgOps.cpp:189]: I think we want to handle the case that not iterates on innermost loop. Eg,

```
for i = ...
  for j = ...
     %0 = load %arg0[i] : memref<100xf32>
     %1 = load %arg1[i, j] : memref<100x100xf32>
     ...
```

something could be from broadcast_add.

I just found that this is not the case in our E2E tests. But I think it is still valid if we broadcast to different dim.

In Linalg form, can be something like

```
func @broadcast_add() {
  %0 = iree.placeholder for "interface buffer" {binding = @legacy_io::@arg0} : memref<4xf32>
  %1 = iree.placeholder for "interface buffer" {binding = @legacy_io::@arg1} : memref<4x4xf32>
  %2 = iree.placeholder for "interface buffer" {binding = @legacy_io::@ret0} : memref<4x4xf32>
  linalg.generic {args_in = 2 : i64,
                  args_out = 1 : i64,
                  indexing_maps = [affine_map<(d0, d1) -> (d0)>,         <---------- here's the case
                                   affine_map<(d0, d1) -> (d0, d1)>,
                                   affine_map<(d0, d1) -> (d0, d1)>],
                  iterator_types = ["parallel", "parallel"]
  } %0, %1, %2 {
  ^bb0(%arg0: f32, %arg1: f32, %arg2: f32):  // no predecessors
    %3 = addf %arg0, %arg1 : f32
    linalg.yield %3 : f32
  }: memref<4xf32>, memref<4x4xf32>, memref<4x4xf32>
  return
}
```

We can also relax the condition later if you think it's better. What do you think?

File: iree/compiler/Conversion/LinalgToVector/VectorizeLinalgOps.cpp
Comment (asaadaldien) [reply iree/compiler/Conversion/LinalgToVector/VectorizeLinalgOps.cpp:173]: mmm, this looks like broadcast getting into its own dispatch region, it should disappear after this https://github.com/google/iree/commit/15c6d4565b1590ed2fcccb1056db7f7478b95101, no ?

File: iree/compiler/Conversion/LinalgToVector/VectorizeLinalgOps.cpp
Comment (hanhanW) [reply iree/compiler/Conversion/LinalgToVector/VectorizeLinalgOps.cpp:201]: There are several checks to decide if we wanna vectorize the Linalg op. If we handle it in a separately pattern, it seems that we need to verify the same thing for several times (O(n) times where n is the number of operands of Linalg op). 

I think in the end we don't need to rewrite the pattern. A better idea is to use vector.type_cast on the operands. But this would need more and more works. I think the conversion from vector to other dialects (eg SPIR-V) is not implemented yet.

File: iree/compiler/Conversion/LinalgToVector/VectorizeLinalgOps.cpp
Comment (hanhanW) [reply iree/compiler/Conversion/LinalgToVector/VectorizeLinalgOps.cpp:33]: Sounds good, I think `mlir/Dialect/Vector/VectorUtils.h` would be the right place. Let's do the cleanup later since it's another repo?

File: iree/compiler/Conversion/LinalgToVector/VectorizeLinalgOps.cpp
Comment (hanhanW) [reply iree/compiler/Conversion/LinalgToVector/VectorizeLinalgOps.cpp:173]: oh yes, it's gone. I observed it in MobileNetV2 and it's working without this now. thanks!

File: iree/compiler/Conversion/LinalgToVector/VectorizeLinalgOps.cpp
Comment (hanhanW) [reply iree/compiler/Conversion/LinalgToVector/VectorizeLinalgOps.cpp:183]: After thinking a while, I think we can relax the condition later.

**Comment by hanhanW** [comment]
I implemented the last piece we need and found that someone just commit it...
https://reviews.llvm.org/rGef7a255c037ca462f71ddd3d2b5a46310b08f6eb

This hasn't integrated into google branch and main branch. Once it's landed, I think we can use the vector pass by default. All the tests on OSS side are passing.

**Comment by hanhanW** [comment]
> I implemented the last piece we need and found that someone just commit it...
> https://reviews.llvm.org/rGef7a255c037ca462f71ddd3d2b5a46310b08f6eb
> 
> This hasn't integrated into google branch and main branch. Once it's landed, I think we can use the vector pass by default. All the tests on OSS side are passing.

It would fail on model integration tests. To pass MobileNetV*, we need to add a canonicalization pattern on `mhlo.sqrt` (I have a pending change in local). For ResNet50, I'm still investigating the issue.

File: iree/compiler/Conversion/LinalgToVector/VectorizeLinalgOps.cpp
Comment (hanhanW) [reply iree/compiler/Conversion/LinalgToVector/VectorizeLinalgOps.cpp:153]: You are right on this, thanks for catching this!

File: iree/compiler/Conversion/LinalgToVector/VectorizeLinalgOps.cpp
Comment (hanhanW) [reply iree/compiler/Conversion/LinalgToVector/VectorizeLinalgOps.cpp:60]: Good catch, let's only consider scalars.

File: iree/compiler/Conversion/LinalgToVector/VectorizeLinalgOps.cpp
Comment (hanhanW) [reply iree/compiler/Conversion/LinalgToVector/VectorizeLinalgOps.cpp:47]: Done

File: iree/compiler/Conversion/LinalgToVector/Passes.h
Comment (hanhanW) [reply iree/compiler/Conversion/LinalgToVector/Passes.h]: done, thanks!

**Review by ThomasRaoux — APPROVED** [review]

File: iree/compiler/Conversion/LinalgToVector/Passes.h
Comment (MaheshRavishankar) [inline iree/compiler/Conversion/LinalgToVector/Passes.h:24]: Could we change the name of the pass to createLoadStoreVectorizationPass(). Easy to confuse that this is going from Linalg -> Vector dialect, which it is not.

File: iree/compiler/Conversion/LinalgToVector/VectorizeLinalgOps.cpp
Comment (MaheshRavishankar) [reply iree/compiler/Conversion/LinalgToVector/VectorizeLinalgOps.cpp:33]: Ok, we have a bunch of things like this. Will have to track it in an issue.

File: iree/compiler/Conversion/LinalgToVector/VectorizeLinalgOps.cpp
Comment (MaheshRavishankar) [inline iree/compiler/Conversion/LinalgToVector/VectorizeLinalgOps.cpp:56]: Please add documentation comment.

File: iree/compiler/Conversion/LinalgToVector/VectorizeLinalgOps.cpp
Comment (MaheshRavishankar) [inline iree/compiler/Conversion/LinalgToVector/VectorizeLinalgOps.cpp:70]: Please mark `static` and add documentation comment.

File: iree/compiler/Conversion/LinalgToVector/VectorizeLinalgOps.cpp
Comment (MaheshRavishankar) [inline iree/compiler/Conversion/LinalgToVector/VectorizeLinalgOps.cpp:91]: Documentation comment here and below.

File: iree/compiler/Conversion/LinalgToVector/VectorizeLinalgOps.cpp
Comment (MaheshRavishankar) [inline iree/compiler/Conversion/LinalgToVector/VectorizeLinalgOps.cpp:76]: Dont know what the convention is but for `Value` return type I usually use `return nullptr`

File: iree/compiler/Conversion/LinalgToVector/VectorizeLinalgOps.cpp
Comment (MaheshRavishankar) [inline iree/compiler/Conversion/LinalgToVector/VectorizeLinalgOps.cpp:121]: This is done in every class. Maybe have a base `VectorizeOpBase` that checks this and calls an apply method to do the vectorization.

It would have been really nice to use [this](https://stackoverflow.com/questions/50754525/pass-an-array-variable-as-template-argument) trick to have a single implementation for all these classes, and just set the arguments to vectorize during instantiation. Seems like a C++-17 feature though.

File: iree/compiler/Conversion/LinalgToVector/VectorizeLinalgOps.cpp
Comment (MaheshRavishankar) [reply iree/compiler/Conversion/LinalgToVector/VectorizeLinalgOps.cpp:180]: Makes sense. Thanks for explaining.

**Review by MaheshRavishankar — APPROVED** [review]
Nice! Just a few nits. Approving assuming they will be addressed.

File: iree/compiler/Conversion/LinalgToVector/Passes.h
Comment (hanhanW) [reply iree/compiler/Conversion/LinalgToVector/Passes.h:24]: Done

File: iree/compiler/Conversion/LinalgToVector/VectorizeLinalgOps.cpp
Comment (hanhanW) [reply iree/compiler/Conversion/LinalgToVector/VectorizeLinalgOps.cpp]: Sure, let's follow your convention.

File: iree/compiler/Conversion/LinalgToVector/VectorizeLinalgOps.cpp
Comment (hanhanW) [reply iree/compiler/Conversion/LinalgToVector/VectorizeLinalgOps.cpp]: Sounds good, thanks!

File: iree/compiler/Conversion/LinalgToVector/LoadStoreVectorization.cpp
Comment (ScottTodd) [inline iree/compiler/Conversion/LinalgToVector/LoadStoreVectorization.cpp:33]: clang-tidy is complaining that this `getBitWidth` function is not used.

Comment (hanhanW) [reply iree/compiler/Conversion/LinalgToVector/LoadStoreVectorization.cpp:33]: Oh, seems that it's gone while refactoring. I'll send out a clean up PR. Thanks!

**Pull request merged**
